### ISSUE TITLE
Parameter controls not working

### ISSUE THREAD

**Issue by jacob-carlborg-apoex** [ViewComponent/lookbook#233]
### Describe the bug

I've integrated Lookbook to our application but some of the parameter controls are not working. The once I've tested which are not working are: `toggle` and `select`. Nothing happens when I click on a toggle or a select box.

### To Reproduce

1. Go to the Lookbook page of the application
2. Select a component
3. Click on the `Params` tab 
4. Click on the toggle for a parameter
5. Nothing happens and no errors in the browser console

### Expected behavior

I expect the toggle to change state and the preview to be updated.

### Version numbers 

* Lookbook: 1.3.3
* ViewComponent: 2.74.1
* Rails: 6.1.5.1
* Ruby: 2.7.2p137 (2020-10-01 revision 5445e04352) [arm64-darwin20]
* view_component-contrib: 0.1.1
* Safari: 16.1 (18614.2.9.1.12)
* Chrome: 107.0.5304.87 (Official Build) (arm64)

### Additional context

It seems to be some issue with the JavaScript, but there are no errors in the browser console. The tree that lists the components are working properly so the JavaScript is not completely broken. If I change the parameter type to `text` everything works fine and the preview is automatically updated when I change the parameter value.

I've also tested to generate a new Rails application with ViewComponent and Lookbook and then everything works as expected. So there's something weird with the setup/configuration in our application. I've tried to remove gems from our application until only the standard Rails gems (those that are included when generating a new Rails application) are left plus ViewComponent and Lookbook, more or less. It didn't help.

I've included the debug data from Lookbook below. Note, `config.components_path` is not correct. We have our components in `app/frontend/components` and `app/frontend/pages`.

```json
{
  "version": "1.3.3",
  "env": "development",
  "config": {
    "project_name": "Lookbook",
    "log_level": 2,
    "log_use_rails_logger": true,
    "auto_refresh": true,
    "components_path": "app/components",
    "page_controller": "Lookbook::PageController",
    "page_route": "pages",
    "page_paths": [
      "test/components/docs"
    ],
    "page_extensions": [
      "html.*",
      "md.*"
    ],
    "page_options": {
    },
    "markdown_options": {
      "tables": true,
      "fenced_code_blocks": true,
      "disable_indented_code_blocks": true,
      "strikethrough": true,
      "highlight": true,
      "with_toc_data": true,
      "lax_spacing": true
    },
    "highlighter_options": {
      "theme": "github",
      "dark": false
    },
    "sort_examples": false,
    "preview_paths": [
      "/Users/jacobcarlborg/.rvm/gems/ruby-2.7.2@my-rails-app/gems/view_component-contrib-0.1.1/app/views",
      "/Users/jacobcarlborg/development/apoex/my-rails-app/app/frontend/components",
      "/Users/jacobcarlborg/development/apoex/my-rails-app/app/frontend/pages"
    ],
    "preview_display_options": {
    },
    "preview_disable_action_view_annotations": true,
    "preview_params_options_eval": false,
    "listen": true,
    "listen_paths": [
      "/Users/jacobcarlborg/.rvm/gems/ruby-2.7.2@my-rails-app/gems/view_component-contrib-0.1.1/app/views",
      "/Users/jacobcarlborg/development/apoex/my-rails-app/app/frontend/components",
      "/Users/jacobcarlborg/development/apoex/my-rails-app/app/frontend/pages",
      "app/components"
    ],
    "listen_extensions": [
      "rb",
      "html.*"
    ],
    "listen_use_polling": false,
    "ui_theme": "zinc",
    "ui_theme_overrides": {
    },
    "ui_favicon": true,
    "debug_menu": true,
    "experimental_features": false,
    "preview_controller": "ViewComponentsController",
    "panels": {
      "main": [
        {
          "label": "Preview",
          "hotkey": "v",
          "disabled": false,
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/inspector/panels/preview",
          "name": "preview",
          "group": "main"
        },
        {
          "label": "HTML",
          "hotkey": "h",
          "disabled": false,
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/inspector/panels/output",
          "name": "output",
          "group": "main"
        }
      ],
      "drawer": [
        {
          "label": "Source",
          "hotkey": "s",
          "disabled": false,
          "show": true,
          "copy": "#<Proc:0x0000000126d74200 (eval):1 (lambda)>",
          "locals": {
          },
          "partial": "lookbook/inspector/panels/source",
          "name": "source",
          "group": "drawer"
        },
        {
          "label": "Notes",
          "hotkey": "n",
          "disabled": "#<Proc:0x0000000126d6e030 (eval):1 (lambda)>",
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/inspector/panels/notes",
          "name": "notes",
          "group": "drawer"
        },
        {
          "label": "Params",
          "hotkey": "p",
          "disabled": "#<Proc:0x0000000126d674d8 (eval):1 (lambda)>",
          "show": true,
          "copy": null,
          "locals": {
          },
          "partial": "lookbook/inspector/panels/params",
          "name": "params",
          "group": "drawer"
        }
      ]
    },
    "inputs": {
      "select": {
        "name": "select",
        "partial": "lookbook/inspector/inputs/select",
        "options": {
          "opts": {
          }
        }
      },
      "textarea": {
        "name": "textarea",
        "partial": "lookbook/inspector/inputs/textarea",
        "options": {
          "opts": {
          }
        }
      },
      "toggle": {
        "name": "toggle",
        "partial": "lookbook/inspector/inputs/toggle",
        "options": {
          "opts": {
          }
        }
      },
      "color": {
        "name": "color",
        "partial": "lookbook/inspector/inputs/color",
        "options": {
          "opts": {
          }
        }
      },
      "range": {
        "name": "range",
        "partial": "lookbook/inspector/inputs/range",
        "options": {
          "opts": {
          }
        }
      },
      "text": {
        "name": "text",
        "partial": "lookbook/inspector/inputs/text",
        "options": {
          "opts": {
          }
        }
      },
      "email": {
        "name": "email",
        "partial": "lookbook/inspector/inputs/text",
        "options": {
          "opts": {
          }
        }
      },
      "number": {
        "name": "number",
        "partial": "lookbook/inspector/inputs/text",
        "options": {
          "opts": {
          }
        }
      },
      "tel": {
        "name": "tel",
        "partial": "lookbook/inspector/inputs/text",
        "options": {
          "opts": {
          }
        }
      },
      "url": {
        "name": "url",
        "partial": "lookbook/inspector/inputs/text",
        "options": {
          "opts": {
          }
        }
      },
      "date": {
        "name": "date",
        "partial": "lookbook/inspector/inputs/text",
        "options": {
          "opts": {
          }
        }
      },
      "datetime_local": {
        "name": "datetime_local",
        "partial": "lookbook/inspector/inputs/text",
        "options": {
          "opts": {
          }
        }
      }
    },
    "tags": {
      "hidden": {
        "name": "hidden",
        "label": "Hidden",
        "options": {
          "label": "Hidden status",
          "opts": {
          }
        }
      },
      "label": {
        "name": "label",
        "label": "Label",
        "options": {
          "label": "Label",
          "opts": {
          }
        }
      },
      "display": {
        "name": "display",
        "label": "Display",
        "options": {
          "label": "Display",
          "opts": {
          }
        }
      },
      "position": {
        "name": "position",
        "label": "Position",
        "options": {
          "label": "Position",
          "opts": {
          }
        }
      },
      "id": {
        "name": "id",
        "label": "Id",
        "options": {
          "label": "Id",
          "opts": {
          }
        }
      },
      "component": {
        "name": "component",
        "label": "Component",
        "options": {
          "label": "Component",
          "opts": {
          }
        }
      },
      "param": {
        "name": "param",
        "label": "Param",
        "options": {
          "label": "Param",
          "opts": {
          }
        }
      },
      "logical_path": {
        "name": "logical_path",
        "label": "Logical Path",
        "options": {
          "label": "Logical Path",
          "opts": {
          }
        }
      }
    }
  }
}
```

Below is a skeleton of a component that reproduces the issue.

Preview:

```ruby
class Flowbite::Button::Preview < ViewComponent::Preview
  # @param foo [Boolean] toggle foobar
  def button_primary(foo: false)
    render(Flowbite::Button::Component.new(foo, "#"))
  end
end
```

Component:

```ruby
class Flowbite::Button::Component < ViewComponent::Base
  attr_reader :name
  attr_reader :options

  def initialize(name = nil, options = nil)
    @name = name
    @options = options
  end
end
```

Template:

```haml
%flowbite-button= link_to(name, options)
```

File structure:

```
app/frontend/
├── components
│   ├── flowbite
│   │   ├── button
│   │   │   ├── component.html.haml
│   │   │   ├── component.rb
│   │   │   └── preview.rb
```

**Comment by jacob-carlborg-apoex** [comment]
If someone could point me to the code that should be run when a toggle is clicked, I can try to debug the issue on my side.

**Comment by allmarkedup** [comment]
hey @jacob-carlborg-apoex, thanks for all the detailed info and I'm really sorry for taking so long to get back to you. I've had some urgent family issues that I've had to take time out for so haven't had a chance to try and reproduce this yet I'm afraid 😕

You are welcome to poke about in the code but I haven't yet got much documentation together on rebuilding assets etc yet so it may be harder work than it should be. In terms of relevant code, these are the main files of interest:

* [Template that renders the toggle](https://github.com/ViewComponent/lookbook/blob/main/app/views/lookbook/inspector/inputs/_toggle.html.erb)
* [JS that handles updating the URL](https://github.com/ViewComponent/lookbook/blob/main/app/assets/lookbook/js/components/params_input.js) 

Lookbook uses AlpineJS under the hood. The toggle has a `click` handler bound to it that updates the value of the `value` property, which is being watched for changes. When a value change happens, the URL is updated and the page is refreshed. 

Looking at the code now one thing that could possibly be happening is that the form validity check is failing on those inputs for some reason. I've just pushed up a branch that removes that check, could you give it a test and see if it changes anything for you?

```rb
gem "lookbook", git: "git://github.com/ViewComponent/lookbook.git", branch: "remove-params-validity-check"
```

If that doesn't help at all let me know and when I have caught up with myself I'll try and reproduce it using your example above. I may still be a little slow at replying for the next few days but hopefully we can get this sorted out for you soon.

**Comment by jacob-carlborg-apoex** [comment]
> Looking at the code now one thing that could possibly be happening is that the form validity check is failing on those inputs for some reason. I've just pushed up a branch that removes that check, could you give it a test and see if it changes anything for you?

That didn't help. I was one patch version behind, which didn't seem to have the validations.

I put a breakpoint in the [`init`](https://github.com/ViewComponent/lookbook/blob/d4c53fef3a833174c62b13cecb55735bf82b69ab/app/assets/lookbook/js/components/params_input.js?rgh-link-date=2022-11-17T20%3A43%3A26Z#L7) and [`update`](https://github.com/ViewComponent/lookbook/blob/d4c53fef3a833174c62b13cecb55735bf82b69ab/app/assets/lookbook/js/components/params_input.js?rgh-link-date=2022-11-17T20%3A43%3A26Z#L11) methods. The browser hits the breakpoint in the `init` method but not the one in the `update` method.

**Comment by allmarkedup** [comment]
@jacob-carlborg-apoex okay thanks, it's a really odd one! Will try to replicate in a fresh app using your setup info above.

One more question - when you click the toggle, does the URL in the browser URL bar change? Or does nothing at all happen? I'm assuming the latter if the update method isn't getting run but just want to double check.

**Comment by jacob-carlborg-apoex** [comment]
> One more question - when you click the toggle, does the URL in the browser URL bar change?

No, the URL doesn't change.

**Comment by allmarkedup** [comment]
@jacob-carlborg-apoex I'm afraid I cannot replicate this - I've created a fresh Rails app based on the information you provided above, with the same component and preview and all seems to work fine for me.

I'm really sorry! I wish I could help get to the bottom of this for you but unless I replicate the issue for myself I don't think there is much else I can do to make progress on this 😕

**Comment by jacob-carlborg-apoex** [comment]
No worries, I had a feeling from the beginning it would be difficult to reproduce. It's the same with all our components.

@allmarkedup it's the `init` method that setups listening on events and `this.$watch` is part of the AlpineJS? Is that correct? Perhaps it's something in AlpineJS that is broken with out setup.

**Comment by allmarkedup** [comment]
@jacob-carlborg-apoex yes there is an AlpineJS `@click` handler in the markup that updates the value of the `value` prop when triggered. The (AlpineJS) `this.$watch` statement basically just watches our for changes to the `value` property and then runs the `update` method. Or it should do anyway - but somehow it seems like either the `@click` handler is not firing correctly or the `$watch` is not registering the change to the property.

Without being able to actually dig into it myself I'm not sure how much progress I can make on this now though I'm afraid.

I'm just clutching at straws here but if you remove all JS from your component preview layout template (or application layout template, whatever you are using to render the components in) does that make any difference? I should be possible for there to be leakage of JS between your previews and the Lookbook UI but it may somehow be happening and clashing? Pretty long shot though really.

If you manage to replicate it somehow on a fresh example rails app then I'd be very happy to look at it, but otherwise I'm afraid I think I've run out of ideas here, unless you manage to dig up any other clues. Sorry I couldn't be of more help in the end.

**Comment by jacob-carlborg-apoex** [comment]
@allmarkedup I've managed to replicate it. It's not a completely fresh Rails application, but very close to. It's available here [1]. Note, for some reason I had to add this line: https://github.com/jacob-carlborg-apoex/lookbook-toggle-test/blob/8e0b1e73ee5ac4e42bbe36ce7a97dc5ef4a514d9/config/application.rb#L14.

[1] https://github.com/jacob-carlborg-apoex/lookbook-toggle-test

**Comment by allmarkedup** [comment]
@jacob-carlborg-apoex this is great, thank you! Really appreciate you taking the time to put this together. I've just pulled it down and I can also replicate the problem which is a game-changer :-)

I'm going to dig into this now. I'll let you know when I have anything worth reporting. Thanks again!

**Comment by allmarkedup** [comment]
@jacob-carlborg-apoex okay I have a fix although it's definitely a bit of an odd one... for some reason the attribute keys on that button element are being escaped in a way that the `@` symbol in the AlpineJS `@click` handler is being replaced by an underscore. So Alpine is never binding the click even on there at all.

Quite why it's just that one element I have no idea about - I have used `@click` handlers thoughout the component and view templates and everywhere else seems fine. And even the toggle works fine on my apps... it's very odd.

The good thing is that it's an easy fix now that I know what it is! ☺️ I'm just getting a PR together with a fix in and I'll get it out in the next release, hopefully over the weekend.

Thanks again for putting the effort in to helping figure this out, I'm sure you are not the only one running into it so it's greaet to get it fixed.

**Comment by jacob-carlborg-apoex** [comment]
Awesome, looking forward to try the fix.

**Comment by allmarkedup** [comment]
@jacob-carlborg-apoex I've just (finally!) released v1.4.2 that should fix the problems you have been running into. I've tested it with both `toggle` and `select` param field types within the context of the repo you provided and both are working for me now.

I'm going to close this issue now but if you are still running into problems after updating please do reopen it. And thanks for your patience with this, it has taken much longer to get this fix out in a release than I would have normally liked!

**Comment by jacob-carlborg-apoex** [comment]
@allmarkedup I've tried with version 1.4.2 and can confirm that both toggles and selects are working. I see now that selects actually did work before. I think I passed the arguments to `@param` in the wrong order before. It rendered a select element but without any options inside it. I assumed it was dependent on JavaScript just as the toggle but it probably isn't. My misstake.

> And thanks for your patience with this, it has taken much longer to get this fix out in a release than I would have normally liked!

No worries. Thank you for tracking this down and fixing the bug, I'm just glad it was fixed 😄.
